### DISCUSSION TITLE
Mouse keys on input overlay

### DISCUSSION THREAD

**Discussion by l3lackShark** [ppy/osu#13043] [Ideas]
**Describe the missing feature:** 
Currently, lazer merges input from the mouse and the keyboard. I think that it's important to have mouse keys displayed separately. For simplicity and design reasons we don't even need a second overlay, just glowing the keys with different color would be more than enough.  

**Proposal designs of the feature:** 
https://www.youtube.com/watch?v=4whB6MRANw4

**Comment by peppy** [comment]
How does this work with joystick inputs? midi inputs? is each category going to be separated? I'm not sure I see the point.

**Reply by Doksha** [reply]
This seems like a strange argument, the minority of players you are trying to cater to (joystick/midi) is so much smaller than the "minority" of players who play with anything other than just two keyboard keys. most players try a playstyle like that at least once, while the quantity of players who would even try joystick/midi in standard is in the single digits.
The overlay would still work as it does now with joystick/midi

**Reply by frenzibyte** [reply]
> The overlay would still work as it does now with joystick/midi

This won't make any sense to the end user if you're already displaying a separate section that's dedicated to mouse keys.

**Reply by Doksha** [reply]
this would only be a toggle in settings, with it enabled you would see all the keys as separate if compatible, but otherwise the two pairs would be combined.
Could you please elaborate what you mean by joystick/midi being different?
I dont have either of those so I dont know how lazer handles them, but I would think they are treated the same as keyboard inputs.?
whatever key is bound to K1 (the rightward key option) would display as K1, whatever key is bound to M1 (the leftward key option) would display as M1, how would that be different for midi/joystick?

**Reply by frenzibyte** [reply]
> Could you please elaborate what you mean by joystick/midi being different?
I dont have either of those so I dont know how lazer handles them, but I would think they are treated the same as keyboard inputs.?
whatever key is bound to K1 (the rightward key option) would display as K1, whatever key is bound to M1 (the leftward key option) would display as M1, how would that be different for midi/joystick?

Are you saying that joystick/midi input should increment K1/K2 while mouse input would increment M1/M2? You understand that doesn't make sense to the user, right?

Because two reasons, 1. why would mouse display separately but joystick/midi remain grouped with keyboard. 2. joystick buttons aren't really "keys" to be grouped with the keyboard side of the key overlay?

**Reply by Doksha** [reply]
I have no idea what you are trying to say... Currently midi/joystick is treated identical to keyboard/mouse.
for example lets say I bind joystick 1 ( X ) to left button (left), joystick 3 ( B ) to right button (left), keyboard z ( z ) to left button (right), and keyboard x ( x ) to right button (right)

Then, in game I hit the sequence X B x B z
that would translate to K1 K2 M1 K2 M2
there is no issue here.

it seems you want joystick to be a mouse key... so why not bind it that? joystick isnt bound to anything by default anyway (I havent checked with midi though)

> Are you saying that joystick/midi input should increment K1/K2 while mouse input would increment M1/M2? You understand that doesn't make sense to the user, right?

I have no idea where you got that idea

>Because two reasons, 1. why would mouse display separately but joystick/midi remain grouped with keyboard. 2. joystick buttons aren't really "keys" to be grouped with the keyboard side of the key overlay?

do you even know how the binding system works? seems like such a strange argument when you could easily go in game and check what you are saying before posting (unless you dont have a joystick/midi)


In case what I am saying is not clear

there are 4 key options, 2 are left button, and 2 are right button. one of each pair is "Mouse" and one of each pair is "Keyboard" whether these are actually bound to their respective options is up to the user.
The proposal is for these separate options to also affect how keys are displayed in game (so that there are up to 4 tapping keys recorded.) whether through color on 2 keys (imo this is a bad idea) or 4 separate individual keys (or both; chosen through skinning)
Or even none at all, but still recording the separate inputs as separate to the replay, it wouldnt affect the user at all, but would be a nice addition for watching replays in stable/danser.

**Reply by ghost** [reply]
ok but how does the code know which key is considered the mouse button and which key is considered the keyboard key

as far as i know the game makes no distinction from them and considers them the same button, thus the current problem

**Reply by Doksha** [reply]
and thats why I am suggesting they are separated (context may have been lost since I also spoke under a duplicate post)

The change that needs to be made is the separation between the "mouse" and  the "keyboard" sides of the input.

**Reply by frenzibyte** [reply]
>>  Are you saying that joystick/midi input should increment K1/K2 while mouse input would increment M1/M2? You understand that doesn't make sense to the user, right?
> I have no idea where you got that idea

From your [original comment](https://github.com/ppy/osu/discussions/13043#discussioncomment-4089981): "The overlay would still work as it does now with joystick/midi"

> there are 4 key options, 2 are left button, and 2 are right button. one of each pair is "Mouse" and one of each pair is "Keyboard" whether these are actually bound to their respective options is up to the user.

This is precisely why I say this doesn't make sense. Allowing the user to set any of the buttons that belong to the "Mouse" overlay to non-mouse keys is straight up dodgy. And even if better names were given, the effort required to support it isn't worth anyone's time unless there's a high demand on it.

>> Because two reasons, 1. why would mouse display separately but joystick/midi remain grouped with keyboard. 2. joystick buttons aren't really "keys" to be grouped with the keyboard side of the key overlay?
> do you even know how the binding system works? seems like such a strange argument when you could easily go in game and check what you are saying before posting (unless you dont have a joystick/midi)

I'm not sure I understand what you're trying to point at. I'm pointing out reasons as to why it wouldn't make sense, what does that have anything to do with the "binding system" or checking the game for anything?

I suggest you tone your attitude down and leave this thread behind as I don't think you're going anywhere here other than repeating yourself.

**Reply by Doksha** [reply]
You implied that somehow joystick/midi were already bound by default (meaning the user doesnt need to bind the buttons themselves, and that they MUST be bound to keyboard) this is not the case, you need to bind the buttons, therefore you have all the choice you need when it comes to differentiating the 4 keys.

Sorry if my statement came off as rude, it was hard for me to understand your argument when most statements seemed to contradict themselves with the game.

**Comment by bdach** [comment]
What's the reasoning for needing this in the first place? Why does it matter?

**Reply by nagi-desuuu** [reply]
I'm guessing that it's for feature parity with stable, and that a lot of skins have the legacy input indicator skinned to match the overall theming of the skin.

**Reply by Doksha** [reply]
because a large number of people who use this feature in stable, its even more useful in lazer since you can bind other buttons than M1 and M2 to the mouse button side. Preventing these from showing up does literally nothing positive, while having them would have no negatives.
The only arguments against it arent against it, just some strange leaps of logic that assign no negatives to the functionality.

**Reply by NotGumballer91** [reply]
A better title would be: Implement legacy input overlay

**Reply by bdach** [reply]
you mean like https://github.com/ppy/osu/issues/10944.

i'm not actually super sure what to do with that thing. half inclined to say that the k1/k2/m1/m2 split may not be supported anymore.

**Reply by Velyro** [reply]
For people who play with 4 keys, it would be nice to check your replays to see if you're actually playing 4 key and not just reverting to alternating, because it can be hard to tell sometimes when you're so used to one thing. 
That's why I personally think this is a good idea, to have 2 of each left and right buttons be displayed on the overlay if you use them, but no need to include 4 if you don't. 
But I dunno, maybe it's not worth the bother if not enough people play 4 key, might just have to see if I can learn how to implement it myself

**Comment by tadatomix** [comment]
I know some of y'all might not want to continue this discussion, but here's the idea. LegacyCounter will have all of the buttons (Could be called B1(Button 1), B2 (Button 2), B1A(Button 1 ALT), B2A (Button 2 ALT)) and they will behave like in stable. So when K1 is pressed, it'll be counted as B1, when M1 is pressed, it'll be counted as B1A. That way, I think, we wouldn't touch existing modern counters (Argon, Triangles) and we will have what people are asking for

**Reply by tadatomix** [reply]
Reason why I'm offering this idea is because of this weird gap in legacy skins. It shouldn't be here. 

![osu!_n62QHvcf9A](https://github.com/user-attachments/assets/16af90b4-452c-4afa-8d4c-d998447c61e1)

**Comment by ckhrix** [comment]
Can we revive this issue?
We can bind 4 buttons in osu standard, why cant we see a counter for each? This is really annoying for us 4 key players, and any other type of alternate input players like tapx etc.

**Comment by cl8n** [comment]
I didn't see anyone in this thread directly mention yet, the main reason I want this is that it gives some insight into a player's tapping style when you view their replays. playstyles like mouse-only or tapx leave behind a signature of sorts on the key counter that is unfortunately lost in lazer. from what I can tell, this information isn't even being stored to new lazer replays at the moment

pishifat mentioned this in an update video when the legacy key counter style was added <https://youtu.be/mlVwPrm0t2E?t=412> (6:52 to 7:06)

**Comment by Flamiii** [comment]
It's worth mentioning that the LegacyKeyOverlay element in lazer already supports every other mode but standard, as it supports the 3 keys for catch, the 4 keys for taiko, and even supports more than 4 keys in mania. 

The majority of players use legacy skins with inputoverlay-background.png files that were specifically designed for 4 keys and it will always look bad and have a gap if there are less than 4 keys. It looks out of place on the osu! "classic" and "retro" default skins that are bundled with the game as well. 99% of players also never use the smoke button (B3) and it has no relevance to gameplay whatsoever, so it shouldn't be on the key overlay. 

As someone who is learning 4 key on standard, the only way for me to look back on my mistakes tapping-wise is to record my gameplay with an external key overlay, because my replays don't separate my keypresses into 4 separate keys even though they're 4 separate keybinds. This would not be an issue on stable since I could just bind my keys to K1, K2, M1, and M2, and then look back at the replay to analyze my keypresses. This is a niche scenario, but at the same time, 4 key support is natively built-in to lazer and can be done without any third-party tools or macros and as rare as it is, there are people out there who play with 3/4 keys or Tap ZX as well.

I think a solution could be to separate Left button and Right button in Input to Left button 1 (LB1), Right button 1 (RB1), Left button 2 (LB2), and Right button 2 (RB2) and have these four keys displayed on the key overlay in that order. At the end of the day, the people in this thread just want basic parity with stable for the key overlay in standard, and I hope this happens eventually.
